### PR TITLE
Parse cache information in smbios-lib

### PR DESCRIPTION
**Pull request by mulark**
Breaking change, but I think we should be providing fully usable values early.

### CODE REVIEW

File: src/structs/types/cache_information.rs [hunk 1]
```diff
@@ -147,6 +155,37 @@ impl Serialize for SMBiosCacheInformation<'_> {
...
+/// # Maximum memory capacity, in kilobytes, for this cache item
+#[derive(Serialize, Debug, PartialEq, Eq)]
+pub enum CacheMemorySize {
+    /// Maximum memory capacity in Kilobytes
+    Kilobytes(u64),
+    /// Use `maximum_cache_size_2` to retrieve the maximum capacity
+    SeeMaximumCacheSize2,
+}
+
+impl From<u16> for CacheMemorySize {
```
Comment (jrgerber) [inline src/structs/types/cache_information.rs:167]: Comments for the review:
Maximum Cache Size and Install Size are both size WORD (u16) and follow the same logic:

Maximum size that can be installed
Bit 15 Granularity
0 – 1K granularity
1 – 64K granularity
Bits 14:0 Max size in given granularity
See 7.8.1.

For Cache sizes greater than 2047 MB, the
Maximum Cache Size field is set to 0xFFFF
and the Maximum Cache Size 2 field is
present, the Granularity bit is set to 1b, and
the size set as required; see 7.8.1.

File: src/structs/types/cache_information.rs [hunk 2]
```diff
@@ -147,6 +155,37 @@ impl Serialize for SMBiosCacheInformation<'_> {
...
+        match raw {
+            0xFFFF => CacheMemorySize::SeeMaximumCacheSize2,
+            // When bit 15 is set, the size units of the remaining bits are 64k
+            _ if raw & 0x8000 == 0x8000 => CacheMemorySize::Kilobytes((raw & 0x7FFF) as u64 * 64),
+            _ => CacheMemorySize::Kilobytes(raw as u64),
+        }
+    }
+}
+
+impl From<u32> for CacheMemorySize {
```
Comment (jrgerber) [inline src/structs/types/cache_information.rs:178]: Maximum Cache Size 2 and Install Size 2 are DWORD (u32) and follow the same logic:

If this field is present, for cache sizes of
2047 MB or smaller the value in the Max
size in given granularity portion of the field
equals the size given in the corresponding
portion of the Maximum Cache Size field,
and the Granularity bit matches the value of
the Granularity bit in the Maximum Cache
Size field.
For Cache sizes greater than 2047 MB, the
Maximum Cache Size field is set to 0xFFFF
and the Maximum Cache Size 2 field is
present, the Granularity bit is set to 1b, and
the size set as required; see 7.8.1.
Bit 31 Granularity
0 – 1K granularity
1 – 64K granularity (always 1b
for cache sizes >2047 MB)
Bits 30:0 Max size in given granularity

File: src/structs/types/cache_information.rs [hunk 3]
```diff
@@ -147,6 +155,37 @@ impl Serialize for SMBiosCacheInformation<'_> {
     }
 }
 
+/// # Maximum memory capacity, in kilobytes, for this cache item
+#[derive(Serialize, Debug, PartialEq, Eq)]
+pub enum CacheMemorySize {
+    /// Maximum memory capacity in Kilobytes
+    Kilobytes(u64),
+    /// Use `maximum_cache_size_2` to retrieve the maximum capacity
+    SeeMaximumCacheSize2,
```
Comment (jrgerber) [inline src/structs/types/cache_information.rs:164]: `SeeMaximumCacheSize2` might also be `InstalledCacheSize2`.  This is not likely to confuse anyone based on the context of when it is used.  Or could go more vague with `SeeCacheSize2` which is no better or worse I think.  Using this library requires a degree of reading the SMBIOS standard.  A mitigation is to add to the comment on line 163 to include "or `installed_cache_size_2`"

**Review by jrgerber — APPROVED** [review]
Great improvement!  Absolutely in line with the overall design intentions of this library.

File: src/structs/types/cache_information.rs (hunk 3 above)
Comment (mulark) [reply src/structs/types/cache_information.rs:164]: Good suggestions I'll clean this up a little.

**Pull request merged**